packages/spacecat-shared-tokowaka-client/src/index.js

```diff
@@ -0,0 +1,447 @@
+      const existingOptimization = mergedConfig.tokowakaOptimizations[urlPath];
+
+      if (!existingOptimization) {
+        // URL path doesn't exist in existing config, add it entirely
+        mergedConfig.tokowakaOptimizations[urlPath] = newOptimization;
+        this.log.debug(`Added new URL path: ${urlPath}`);
+      } else {
+        // URL path exists, merge patches
+        const existingPatches = existingOptimization.patches || [];
+        const newPatches = newOptimization.patches || [];
+
+        // Create a map of existing patches by opportunityId+suggestionId
+        const patchMap = new Map();
+        existingPatches.forEach((patch, index) => {
+          const key = `${patch.opportunityId}:${patch.suggestionId}`;
+          patchMap.set(key, { patch, index });
+        });
+
+        // Process new patches
+        const mergedPatches = [...existingPatches];
+        let updateCount = 0;
+        let addCount = 0;
+
+        newPatches.forEach((newPatch) => {
+          const key = `${newPatch.opportunityId}:${newPatch.suggestionId}`;
+          const existing = patchMap.get(key);
+
+          if (existing) {
+            mergedPatches[existing.index] = newPatch;
+            updateCount += 1;
+          } else {
+            mergedPatches.push(newPatch);
+            addCount += 1;
+          }
+        });
+
+        mergedConfig.tokowakaOptimizations[urlPath] = {
+          ...existingOptimization,
+          prerender: newOptimization.prerender,
+          patches: mergedPatches,
+        };
+
+        this.log.debug(`Merged patches for ${urlPath}: ${updateCount} updated, ${addCount} added`);
+      }
+    });
+
+    return mergedConfig;
+  }
+
+  /**
+   * Uploads Tokowaka configuration to S3
+   * @param {string} siteTokowakaKey - Tokowaka API key (used as S3 key prefix)
+   * @param {Object} config - Tokowaka configuration object
+   * @returns {Promise<string>} - S3 key of uploaded config
+   */
+  async uploadConfig(siteTokowakaKey, config) {
+    if (!hasText(siteTokowakaKey)) {
+      throw this.#createError('Tokowaka API key is required', HTTP_BAD_REQUEST);
+    }
+
+    if (!isNonEmptyObject(config)) {
+      throw this.#createError('Config object is required', HTTP_BAD_REQUEST);
+    }
+
+    const s3Path = `opportunities/${siteTokowakaKey}`;
+
+    try {
+      const command = new PutObjectCommand({
+        Bucket: this.bucketName,
+        Key: s3Path,
+        Body: JSON.stringify(config, null, 2),
+        ContentType: 'application/json',
+      });
+
+      await this.s3Client.send(command);
+      this.log.info(`Successfully uploaded Tokowaka config to s3://${this.bucketName}/${s3Path}`);
+
+      return s3Path;
+    } catch (error) {
+      this.log.error(`Failed to upload Tokowaka config to S3: ${error.message}`, error);
+      throw this.#createError(`S3 upload failed: ${error.message}`, HTTP_INTERNAL_SERVER_ERROR);
```

Comment:
for displaying error in UI, we should not highlight internal impl details like S3.